geopandas/tools/hilbert_curve.py

```diff
@@ -0,0 +1,183 @@
+import pandas as pd
+import numpy as np
+
+
+def _hilbert_distance(gdf, total_bounds=None, p=15):
+    """
+    Calculate hilbert distance for a GeoDataFrame
+    int coordinates
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+
+    total_bounds : Total bounds of geometries - array
+
+    p : The number of iterations used in constructing the Hilbert curve
+
+    Returns
+    ---------
+    Pandas Series containing hilbert distances
+
+    """
+    # Calculate bounds as numpy array
+    bounds = gdf.bounds.to_numpy()
+
+    if total_bounds is None:
+        total_bounds = np.array(
+            (
+                np.nanmin(bounds[:, 0]),  # minx
+                np.nanmin(bounds[:, 1]),  # miny
+                np.nanmax(bounds[:, 2]),  # maxx
+                np.nanmax(bounds[:, 3]),  # maxy
+            )
+        )
+
+    # Calculate discrete coords based on total bounds and bounds
+    x, y = _continuous_to_discrete_coords(total_bounds, bounds, p)
+    # Calculate distance from morton curve
+    distances = _encode(p, x, y)
+
+    return pd.Series(distances, index=gdf.index, name="hilbert_distance")
+
+
+def _continuous_to_discrete_coords(total_bounds, bounds, p):
+
+    """
+    Calculates mid points & ranges of geoms and returns
+    as discrete coords
+
+    Parameters
+    ----------
+
+    total_bounds : Total bounds of geometries - array
+
+    bounds : Bounds of each geometry - array
+
+    p : The number of iterations used in constructing the Hilbert curve
+
+    Returns
+    ---------
+    Discrete two-dimensional numpy array
+    Two-dimensional array Array of hilbert distances for each geom
+    """
+
+    # Hilbert Side len
+    side_length = 2 ** p
```

Comment:
Note that in [GEOS this is calculated](https://github.com/libgeos/geos/blob/main/src/shape/fractal/HilbertEncoder.cpp#L32) as `(2 ** p)-1`

Reply:
> Did you notice that when comparing the results you got with GEOS?

Yes, that's how I spotted this.

In the [GEOS test](https://github.com/libgeos/geos/pull/556/commits/72caa8d833901db161a6471ec20e9a4b735bfd49#diff-7067e7dc083a60e06851744edf1d72c9e92b47585c1a941c719eea87f14cb790R61) I was using `POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))` at a level of 2, and in GeoPandas it yields 8 and in GEOS yields 2; when I use `(2 ** p)-1` in GeoPandas they match.

Reply:
I corrected this in https://github.com/geopandas/dask-geopandas/pull/134 (will update here once it is finalized / merged in dask-geopandas)